rfcs/text/0000-matrix-service.md

```diff
@@ -0,0 +1,418 @@
+
+#### Add
+
+The Matrix Service supports additions for new bundles and projects by executing an ETL job for new bundle FQIDs and project UUIDs. Following indexing of a project, a project matrix for each format will be generated via the service and uploaded to the Data Browser-managed `project-assets.data.humancellatlas.org` S3 bucket at the `project-assets/project-matrices/<project_uuid>.<genus_species>.<format>` prefix. This enables the direct download of project matrices via the Data Browser.
+
+Currently, this process initiated manually by DataOps notifying a Matrix Service developer when secondary analysis for a project has completed and the results are validated. A ticket is opened by DataOps in the Matrix Service board to track 1) project ingestion and 2) availability via the Data Browser. These two steps must also be kicked off manually by a Matrix Service developer which requires 1-2 days of engineering attention per project. A [design proposal](https://docs.google.com/document/d/1RCJo_QTeP9dysqrE2JEeI2NH7f5VTLZtB7wnC2kjo1c/edit?usp=sharing) to automate all steps and enable a DataOps or Matrix Service member to kick off the process via an API endpoint is available and [slated for completion by EOQ4 2019](https://app.zenhub.com/workspaces/dcp-5ac7bcf9465cb172b77760d9/issues/humancellatlas/matrix-service/352). When an implementation of project-complete notifications is available, the new endpoint may act as a hook to enable E2E DCP automation.
+
+The team agrees to a 1-week processing period to make available a project following secondary analysis completion with the current manual process. Following automation, this period will be updated to 24 hours to index the project and 48 hours for availability via the Data Browser.
+
+#### Update
+
+The Matrix Service does not respond to bundle updates in real-time as the service does not subscribe to DSS notifications. This is primarily due to the [Matrix cache](#matrix-cache) and the loss in performance when invalidating entries when the index is updated. The cache dramatically expedites the time to receive a matrix, however, not responding to bundle updates risks data consistency across DCP services and data availability to users.
+
+To balance these competing motivations, the team proposes to respond to updates on a monthly cadence in the short term. In the long term, the team aims to support DCP-wide data consistency for versioned releases of HCA data. See [Reproducibility](#reproducibility) for more details on data guarantees.
```

Comment:
It would be good to understand what classes of updates that the matrix services are proposing we conduct on a monthly cadence.

The metadata team is aware that right now we have a higher rate of errors in our metadata that is ideal and this creates churn which impacts all DCP teams. Data Accuracy is a priority for Q1 and we will work with the Data Ops team to create a virtous circle so as new error modes are discovered we can as rapidly as possible update our validation to catch the types of errors that can be caught automatically. This means overtime we should see the rate of errors drop and the frequency by which updates are needed decrease

My concern with only having a monthly cadence for metadata updates is when we have discovered an error in a metadata field which about the impact the scientific interpretability of the data, the easiest of those to understand is if a sample has been labelled with the wrong biological sex. This is a class of error we will never be able to catch automatically but when we are notified of such an error I would want to ensure we are in a position where we can fix it in 24 to 48hrs not 1 month.

Can we work together to understand what the best way forward is here to ensure updates don't create challenges for consumer components but we also don't have to leave erroneous metadata in the system for longer than necessary.

Reply:
Agreed, this section could use more nuance - the updates refer to both inconsequential and consequential metadata updates. Small clarification, updates are proposed to be responded to by the Matrix Service on a monthly cadence as opposed to being conducted by the DCP at this interval. 

The proposed 1-month period can be shortened if insufficient, but can not be real-time to preserve the cache. @lauraclarke Is there a period that would be preferred? With this being said, there are exceptions in which the Matrix Service will expedite and respond immediately to a change such as redaction. However, this should be the exception and not the rule. In the event that an update is critical, the Matrix Service is able to respond to it immediately.

Looking forward, we anticipate with the realization of data releases that both Data Accuracy and Matrix Performance can be achieved by controlling when data is synchronized across the DCP.

Reply:
Thanks for the clarification @calvinnhieu, I had over-interpreted your statement to suggest that updates from a wrangler/ingest perspective shouldn't be pushed on a greater frequency than once a month. 

I agree in the majority of updates a monthly cadence of responsiveness from the matrix service seems reasonable, if we end up in a position where an error gets through which might lead to significant misinterpretation of the data we can review this principle, the simplest solution, in that case, might be to disable the matrix for that project until it is fixed. 

thanks